[JENKINS-38903] Split Exception handling for node provision and adding to Jenkins

pjanouse opened the pull request:
https://issues.jenkins-ci.org/browse/JENKINS-38903

pjanouse committed:
[JENKINS-38903] Split Exception handling for node provision and adding to Jenkins

oleg-nenashev commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -217,29 +218,40 @@ public void run() {
+                                    LOGGER.log(Level.WARNING, "Provisioned agent " + f.displayName + " failed to launch",
+                                            e);
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+                                        cl.onFailure(f, e);
```
If you modify this code, makes sense to make it exception-tolerant. I mean the code should or fail if runtime exception occurs in the listener

stephenc replied:
guard against exceptions thrown here... perhaps the best plan would be to move these loops into a `CloudProvisionlinListener.fireOnFailure(f,e)` and `CloudProvisionlinListener.fireOnComplete(f,node)` static helper methods that can ensure exceptions are not propagated.

Also, we have to ask if it makes sense to fire on complete and then failure where the failure is the IOException saving the node to disk... I wonder should we perform some different behaviour for the failure to save to disk

oleg-nenashev commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -217,29 +218,40 @@ public void run() {
+                                        cl.onComplete(f, node);
+                                    }
+                                } catch (InterruptedException e) {
+                                    throw new AssertionError(e); // since we confirmed that the future is already done
```
NIT: makes sense to add clarification text to the exception

oleg-nenashev reviewed (approved):
LGTM 👍

stephenc commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -217,29 +218,40 @@ public void run() {
+                                Node node = null;
+                                try {
+                                    node = f.future.get();
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
```
I would move this outside of the try catch. That way we do not risk loss of a `Node` instance if a `CloudProvisioningListener` happens to propagate an `ExecutionException`...

stephenc commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -217,29 +218,40 @@ public void run() {
+                                                + "provisioning agent " + f.displayName, cause);
+                                    }
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+                                        cl.onFailure(f, cause);
```
guard against exceptions thrown here

stephenc reviewed (changes requested):
I think there needs to be some fixes on this

pjanouse committed:
Defined new static helper methods that ensure exceptions are not propagated

pjanouse commented:
Guys, I've tried to make a new version based on your feedback.

From my PoV, `onComplete` should be fired when *everything* is in place - it means after successful insert and launch the new node to Jenkins, but I understand something could need to be notified right after the new node is provisioned (before inserting to Jenkins)..

pjanouse commented:
What about a new signal when really everything passes (e.g. either `onConnected()` or `onLive()`)? WDYT?

Such a change can't break a backward-compatibility.

stephenc commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -217,29 +218,43 @@ public void run() {
+                                    } catch (IOException e) {
+                                        LOGGER.log(Level.WARNING, "Provisioned agent " + f.displayName + " failed to launch",
+                                                e);
+                                        for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
```
it would be so much better if the `fireOnFailure` was a static method that wrapped the `for (cl: all()) {` so that you just say fire and it fires it for all

stephenc commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -217,29 +218,43 @@ public void run() {
+
+                                if (node != null) {
+                                    for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+                                        cl.fireOnComplete(f, node);
```
This probably should have been `fireOnProvisioned` then `fireOnComplete` can take place when the node has been saved... the issue as I see it is if plugins are currently using `onComplete` to modify the node before save... given that https://github.com/search?q=org%3Ajenkinsci+CloudProvisioningListener&type=Code doesn't indicate that there are many people using this extension point... I have also analyzed the CloudBees internal usage and none of the 3 closed source implementations we have would object to the splitting.

Thus I suggest that we can change the listener contract and move onCompleted to after the save and introduce onProvisioned here

stephenc replied:
Hmmm thinking about this further... this point is actually the current end-point for the CloudProvisioningListener. It is incorrect to report the IOException from failure to save the node to onFailure... rather we should add two new events, one to capture the node being saved (onCommitted) and one to capture the IOException (onRollback)

pjanouse replied:
Totally agree, if we can change the listener contact, that's the best solution.

stephenc replied:
I think adding new events after the save is the way to go. The original documented contract implies that the existing callbacks will all be only after the `Future` completes, so there is no compatibility issue with adding `onCommitted` (signalling that Jenkins has committed ownership of the node) and `onRollback` (signalling that Jenkins has aborted ownership of the node and the node is now unmanaged) events.

stephenc reviewed (changes requested):
The fire\* methods should handle the for loop.

I have made some suggestions on how to handle the event splitting

pjanouse committed:
Added onCommit and onRollback signals to CloudProvisioningListener

Added the new signals to be able to notify the state after Jenkins.addNode(Node)
All Listener's calls moved to an exception-tolerant static helpers

pjanouse commented:
Rebased and modified.

Note: I've used `onCommit` instead of `onCommitted` because existing `onComplete` (and not `onCompleted`). However there still exists an inconsistency for `onStarted`, but I don't want to break a back-compatibility. If you prefer `onCommitted`, I'l fix it.

oleg-nenashev commented on core/src/main/java/hudson/slaves/CloudProvisioningListener.java:
```diff
@@ -61,20 +64,41 @@ public void onComplete(NodeProvisioner.PlannedNode plannedNode, Node node) {
+     *
+     * @param plannedNode the plannedNode which resulted in the <code>node</code> being provisioned
+     * @param node the node which has been provisioned by the cloud
+     */
```
<code>@since TODO</code>

oleg-nenashev commented on core/src/main/java/hudson/slaves/CloudProvisioningListener.java:
```diff
@@ -61,20 +64,41 @@ public void onComplete(NodeProvisioner.PlannedNode plannedNode, Node node) {
+     * @param node the node which has been provisioned by the cloud
+     */
+    public void onCommit(NodeProvisioner.PlannedNode plannedNode, Node node) {
+
```
Nice to have: Add a comment like "Noop by default". Otherwise tools like Sonar may grumble

oleg-nenashev commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -817,4 +830,65 @@ private static float getFloatSystemProperty(String propName, float defaultValue)
+        for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+            try {
+                cl.onCommit(plannedNode, newNode);
+            } catch (Throwable e) {
```
Throwable suppresses Errors like OutOfMemory exception. These errors are supposed to be fatal in Java.
I would rather handle Exceptions only (FindBugs does not like it IIRC) or propagate Errors

pjanouse replied:
We have 3 options here: 

1. left the original code
.

stephenc replied:
in all these cases you need to rethrow if it is an `Error`

oleg-nenashev reviewed (changes requested):
Needs fix due to Error suppression.
IMHO it would be great to explicitly annotate nullability statuses (Nonnull/CheckForNull) for parameters in all introduced APIs. And <code>@since</code> in Javadoc

pjanouse committed:
Added @Nonnull annotation
Changed the method signature CloudProvisioningListener.onRollback()

pjanouse commented:
Rebased and incorporated Oleg's feedback.

stephenc commented on core/src/main/java/hudson/slaves/CloudProvisioningListener.java:
```diff
@@ -61,20 +65,47 @@ public void onComplete(NodeProvisioner.PlannedNode plannedNode, Node node) {
+     * @since TODO
+     */
+    public void onCommit(@Nonnull NodeProvisioner.PlannedNode plannedNode, @Nonnull Node node) {
+        ; // Noop by default
```
why do we have a `;` here?

stephenc commented on core/src/main/java/hudson/slaves/CloudProvisioningListener.java:
```diff
@@ -61,20 +65,47 @@ public void onComplete(NodeProvisioner.PlannedNode plannedNode, Node node) {
+     */
+    public void onRollback(@Nonnull NodeProvisioner.PlannedNode plannedNode, @Nonnull Node node,
+                           @Nonnull Throwable t) {
+        ; // Noop by default
```
remove redundant `;` it will only cause static analysis to complain

stephenc reviewed (changes requested):
Need to rethrow `Error` instances

pjanouse committed:
Re-throw Error in the fireOnXXX()
Removed re-thrown Throwable in the main try/catch block
	(an instance of the Error is handled separately)

oleg-nenashev commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -817,4 +828,81 @@ private static float getFloatSystemProperty(String propName, float defaultValue)
+                LOGGER.log(Level.SEVERE, "Unexpected uncaught exception encountered while "
+                        + "processing onFailure() listener call in " + cl + " for agent "
+                        + plannedNode.displayName, e);
+                if (e instanceof Error) {
```
🐜 De-facto we process errors in the logging handler. E.g. logging constructs the string, and it may cause OutOfMemoryError, which suppresses the original one. I'd rather vote that Errors are being rethrown without any processing

oleg-nenashev commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -817,4 +828,81 @@ private static float getFloatSystemProperty(String propName, float defaultValue)
+                LOGGER.log(Level.SEVERE, "Unexpected uncaught exception encountered while "
+                        + "processing onRollback() listener call in " + cl + " for agent "
+                        + newNode.getDisplayName(), e);
+                if (e instanceof Error) {
```
same

oleg-nenashev reviewed (changes requested):
Still needs a fix of Error processing

pjanouse committed:
Handling of Error changed

oleg-nenashev commented:
LGTM 👍 
CC @stephenc , who also requested changes

stephenc commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -817,4 +828,77 @@ private static float getFloatSystemProperty(String propName, float defaultValue)
+        for (CloudProvisioningListener cl : CloudProvisioningListener.all()) {
+            try {
+                cl.onFailure(plannedNode, cause);
+            } catch (Throwable e) {
```
`} catch (Error e) { throw E; `

Otherwise the LOGGER.log statement could re-trigger an OOM error)

pjanouse replied:
My oversight, fixing.. sorry'-).

stephenc commented on core/src/main/java/hudson/slaves/NodeProvisioner.java:
```diff
@@ -817,4 +828,77 @@ private static float getFloatSystemProperty(String propName, float defaultValue)
+            try {
+                cl.onRollback(plannedNode, newNode, cause);
+            } catch (Error e) {
+                throw e;
```
+1 this one is done right, vs `fireOnFailure`

stephenc reviewed (changes requested):
Just one try...catch block handling errors the wrong way

pjanouse committed:
Fixed Error instance handling in NodeProvisioner.fireOnFailure()

stephenc reviewed (approved):
:+1:

oleg-nenashev commented:
Merging since there is no other feedback
